tests/nnstreamer_filter_reload/tensor_filter_reload_test.c

```diff
@@ -0,0 +1,279 @@
+ * @author  Dongju Chae <dongju.chae@samsung.com>
+ * @bug     No known bugs except NYI.
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <gst/gst.h>
+#include <tensor_common.h>
+
+#define print_log(...) if (!silent) g_message (__VA_ARGS__)
+#define make_gst_element(element) do{\
+  element = gst_element_factory_make(#element, #element);\
+  if (!element) {\
+    g_printerr ("element %s could not be created.\n", #element);\
+    return_val = -1;\
+    goto out_unref;\
+  }\
+} while (0);
+
+#define IMAGE_FPS (25)
+#define EVENT_INTERVAL (1000)
+#define EVENT_TIMEOUT (10000)
+
+static gboolean silent = TRUE;
+static gchar *input_img_path = NULL;
+static gchar *first_model_path = NULL;
+static gchar *second_model_path = NULL;
+
+static GMainLoop *loop = NULL;
+static gint return_val = 0;
+
+/**
+ * @brief Bus callback function
+ */
+static gboolean
+bus_callback (GstBus * bus, GstMessage * message, gpointer data)
+{
+  print_log ("Got %s message\n", GST_MESSAGE_TYPE_NAME (message));
+
+  switch (GST_MESSAGE_TYPE (message)) {
+    case GST_MESSAGE_ERROR:{
+      GError *err;
+      gchar *debug;
+
+      gst_message_parse_error (message, &err, &debug);
+      print_log ("Error: %s\n", err->message);
+      g_error_free (err);
+      g_free (debug);
+
+      g_main_loop_quit (loop);
+      break;
+    }
+    case GST_MESSAGE_EOS:
+      g_main_loop_quit (loop);
+      break;
+    default:
+      break;
+  }
+
+  return TRUE;
+}
+
+/**
+ * @brief Find the index with a maximum value
+ */
+static gint
+get_maximum_index (guint8 *data, gsize size)
+{
+  gint idx, max_idx = 0;
+  guint8 maximum = 0;
+
+  for (idx = 0; idx < size; ++idx) {
+    if (data[idx] > maximum) {
+      maximum = data[idx];
+      max_idx = idx;
+    }
+  }
+
+  return max_idx;
+}
+
+/**
+ * @brief Signal to handle new output data
+ */
+static GstFlowReturn
+check_output (GstElement *sink, void *data __attribute__((unused)))
+{
+  static gint prev_index = -1;
+
+  GstSample *sample;
+  GstBuffer *buffer;
+  GstMapInfo info;
+  gint index;
+
+  g_signal_emit_by_name (sink, "pull-sample", &sample);
+  if (!sample)
+    return GST_FLOW_ERROR;
+
+  buffer = gst_sample_get_buffer (sample);
+  if (!buffer)
+    return GST_FLOW_ERROR;
+
+  if (!gst_buffer_map (buffer, &info, GST_MAP_READ))
+    return GST_FLOW_ERROR;
+
+  /**
+   * find the maximum entry; this value should be the same with
+   * the previous one even if a model is switched to the other one
+   */
+  index = get_maximum_index (info.data, info.size);
+  if (prev_index != -1 && prev_index != index) {
+    g_critical ("Output is different! %d vs %d\n", prev_index, index);
+    return_val = -1;
+  }
+
+  prev_index = index;
+
+  gst_buffer_unmap (buffer, &info);
+  gst_sample_unref (sample);
+
+  return GST_FLOW_OK;
+}
+
+/**
+ * @brief Reload a tensor filter's model (v1 <-> v2)
+ */
+static gboolean
+reload_model (GstElement *tensor_filter)
+{
+  static gboolean is_first = TRUE;
```

Comment:
Overall LGTM
However, I do not understand the purpose of this variable. As far as I can see, this function is called only once after EVENT_INTERVAL timeout.

Reply:
The callback of timeout is repeatedly called if it returns TRUE. 
The period of timeout is interval between events, not an one-shot trigger.

Reply:
aah right, the timeout is called repeatedly.